expat/doc/reference.html

```diff
@@ -1112,12 +1113,40 @@ <h4 id="XML_Parse">XML_Parse</h4>
 <code>isFinal</code> parameter informs the parser that this is the last
 piece of the document. Frequently, the last piece is empty (i.e.
 <code>len</code> is zero.)
+</p>
+
+<p>
 If a parse error occurred, it returns <code>XML_STATUS_ERROR</code>.
 Otherwise it returns <code>XML_STATUS_OK</code> value.
 Note that regardless of the return value, there is no guarantee that all
 provided input has been parsed; only after <a href="#isFinal">the
 concluding call</a> will all handler callbacks and parsing errors have
 happened.
+</p>
+
+<p>
+Function <code>XML_Parse</code> is mostly a convenience wrapper pairing
+calls to <code><a href="#XML_ParseBuffer">XML_ParseBuffer</a></code>
+and <code><a href="#XML_GetBuffer">XML_GetBuffer</a></code>.
+You can roughly imagine <code>XML_Parse</code>'s implementation as
+a more complete version with error checking of the following:
+</p>
+
+<pre>
+enum XML_Status XMLCALL
+XML_Parse(XML_Parser parser, const char *s, int len, int isFinal) {
+  // Not the actual XML_Parse, oversimplified!
+  void *buff = XML_GetBuffer(parser, len);
+  memcpy(buff, s, len);
+  return XML_ParseBuffer(parser, len, isFinal);
+}</pre>
+
+<p>
+To avoid double copying of the input, direct use of functions
+<code><a href="#XML_ParseBuffer">XML_ParseBuffer</a></code>
+and <code><a href="#XML_GetBuffer">XML_GetBuffer</a></code> is advised
+for most production code.
+</p>
```

Comment:
Maybe we can qualify this to present an easy choice:

> To avoid unnecessary double copying of input:
> * If you're using `read` or similar functionality to fill your buffers, fill directly into the buffer from XML_GetBuffer, then parse with XML_ParseBuffer.
> * If your whole input is already in memory, XML_Parse may be able to parse it all at once without copying into Expat's internal buffer.
> * Otherwise, XML_Parse is functionally equivalent to calling XML_GetBuffer, memcpy, and XML_ParseBuffer.

It's not perfectly formulated; feel free to rephrase or use only parts of my suggestion.

Reply:
Note how it says "may be able" here -- that covers the `XML_CONTEXT_BYTES` condition without going into too much detail. Applications that follow this advice will simply avoid the double copying when XML_CONTEXT_BYTES is unset (or 0, if that change happens), and otherwise be no worse than using Get/ParseBuffer.

Reply:
Let me think about that, this feels right and wrong at the same time. It's too complex.